lib/logstash/inputs/http_poller.rb

```diff
@@ -183,20 +209,69 @@ def setup_schedule(queue)
     schedule_type = @schedule.keys.first
     schedule_value = @schedule[schedule_type]
     raise LogStash::ConfigurationError, msg_invalid_schedule unless %w(cron every at in).include?(schedule_type)
-
     opts = schedule_type == "every" ? { first_in: 0.01 } : {}
     scheduler.public_send(schedule_type, schedule_value, opts) { run_once(queue) }
     scheduler.join
   end
 
+  private
+  def handle_pagination(queue, name, request)
+    method, url, pagination = request
+    pagination = pagination[:pagination]
+    concurrent_requests = !pagination["concurrent_requests"].nil? ? pagination["concurrent_requests"] : 1
+    state_file = pagination["last_run_metadata_path"]
+    current_page, in_progress_pages = @state_handler.start_paginated_request(name, state_file, pagination["start_page"] - 1, 1)
+    in_progress_pages.each do |page|
+      create_paginated_request(queue, name, request, page, pagination)
+    end
+    client.execute!
+
+    while current_page.get < pagination["end_page"] do
+      break if stop?
+      current_page.getAndIncrement
+      @state_handler.add_page(name, current_page)
+      create_paginated_request(queue, name, request, current_page.get, pagination)
+      if @state_handler.in_progress_pages.size >= concurrent_requests
```

Comment:
Since our `StateHandler` is only capable of handling one paginated-request at a time, keeping its "current page" an internal detail would simplify things a bit. 

This section also has a race condition with regard to state persistence. Because we increment the `current_page` _before_ we create the page (which adds it to the `StateHandler`'s `in_progress_pages`), there is a window in which the state file can be written _between_ the two. Resuming from this state file would _skip_ the current page as it would not be listed as in progress.